### PR TITLE
Fix mission check for init mission

### PR DESCRIPTION
**Pull request by KonradRudin**
### Solved Problem
The mission feasibility checks could return a wrong result on startup for the loaded mission on the storage.

### Solution
- Only run the mission feasibility check after the geofence has loaded all the data.
- Only run mission feasibility check after a global position topic has been published.
- Some Bugfixes

### Changelog Entry
For release notes:
```
Bugfix Correclty check mission feasilbity on startup
```

### Alternatives
We really should decouple the feasibility check from the mission execution.

### Test coverage
- SITL test:
  - Test case 1: Define a valid mission and store it on the vehicle. Stop the simulation. Start it again, with a different location far away.
  - Test case 2: Define a mission and a geofence, where the mission violates the geofence. Stop the simulation. Start it again.
  - Test results: Prior in both cases, the vehicle would arm then immediately shut down again since the reevaluation of the mission feasibility returned the expected result. However, with the PR, the vehicle is not even armed, as it properly determines the mission feasibility at startup.

#### BE AWARE OF

- What should be happening if we don't have a global position? (i guess we anyways can't fly the mission, so it should be fine).

### CODE REVIEW

File: src/modules/navigator/mission_base.cpp [hunk 1]
```diff
@@ -1360,3 +1366,12 @@ bool MissionBase::checkMissionDataChanged(mission_s new_mission)
 		(new_mission.mission_id != _mission.mission_id) ||
 		(new_mission.current_seq != _mission.current_seq));
 }
+
+bool MissionBase::canRunMissionFeasibility()
+{
+	return _navigator->home_global_position_valid() && // Need to have a home position checked
+	       _navigator->get_global_position()->timestamp > 0 && // Need to have a position, for first waypoint check
```
Comment (sfuhrer) [inline src/modules/navigator/mission_base.cpp:1373]: Let's check if we can use Home position if on the ground and global pos invalid

**Comment by sfuhrer** [comment]
Okay I've checked the state about requirement when to run the feasibility checker and when not (what is now in `canRunMissionFeasibility()`).
With current main the mission feasibility checks [return false if ho valid Home is set](https://github.com/PX4/PX4-Autopilot/blob/c5835a48deb2abc7d403aaab7a657e1a5f624af8/src/modules/navigator/mission_feasibility_checker.cpp#L71). And it is not possible to set a Home position manually because of it's [rejected if local position is invalid.](https://github.com/PX4/PX4-Autopilot/blob/c5835a48deb2abc7d403aaab7a657e1a5f624af8/src/modules/commander/HomePosition.cpp#L253).
With the changes as proposed here the behavior is different in that it will in the "no GPS" case always upload the mission instead of rejecting it, but then it will check the feasibility of it once it gets a GPS lock. Seems better to me, and it's still safe as you won't be able to execute an invalid mission in both ways.

**Review by sfuhrer — DISMISSED** [review]
All good findings, and for me the right solutions. As you say probably not the last bugs in the logic but certainly a stab in the right direction.

**Comment by jongell** [comment]
Tested with VTOL:

Works with: 
-Existing Takeoff in the mission, but missing landing pattern
-1st Waypoint too far

Found one corner case:
1. MIS_DIST_1WP = 10000
2. Upload normal VTOL Mission with Take off and Land Pattern within 10000m
3. Change MIS_DIST_1WP to 1m
4. Start Mission via Action in AMC

Here the Vehicle first arms, switches to Mission and then triggers the no valid mission failsafe, switches to return, waits and disarms

**Comment by KonradRudin** [comment]
> Tested with VTOL:
> 
> Works with: -Existing Takeoff in the mission, but missing landing pattern -1st Waypoint too far
> 
> Found one corner case:
> 
> 1. MIS_DIST_1WP = 10000
> 2. Upload normal VTOL Mission with Take off and Land Pattern within 10000m
> 3. Change MIS_DIST_1WP to 1m
> 4. Start Mission without reuploading it via Action in AMC
> 
> Here the Vehicle first arms, switches to Mission and then triggers the no valid mission failsafe, switches to return, waits and disarms

Thanks for testing. Yeah, the check is not rerun on parameters change that is why the commander thinks the mission is feasible and switches to the mission. On mission activation however, the check is performed again, and thus, at this point it know that it is infeasible and shuts down again. I think we should address this later when we properly fix/refactor the check. I will only make a small change to make sure that the mission is rechecked on activation (currently depending on timing, you could change the parameter such that the mission is infeasible and still fly the mission). @sfuhrer for visibility

**Review by sfuhrer — APPROVED** [review]

**Pull request merged**